src/passes/OptimizeInstructions.cpp

```diff
@@ -578,10 +579,30 @@ struct OptimizeInstructions
         if (right->type == Type::i32) {
           uint32_t c = right->value.geti32();
           if (IsPowerOf2(c)) {
-            if (binary->op == MulInt32) {
-              return optimizePowerOf2Mul(binary, c);
-            } else if (binary->op == RemUInt32) {
-              return optimizePowerOf2URem(binary, c);
+            switch (binary->op) {
+              case MulInt32:
+                return optimizePowerOf2Mul(binary, c);
+              case RemUInt32:
+                return optimizePowerOf2URem(binary, c);
+              case DivUInt32:
+                return optimizePowerOf2UDiv(binary, c);
+              default:
+                break;
+            }
+          }
+        }
+        if (right->type == Type::i64) {
+          uint64_t c = right->value.geti64();
+          if (IsPowerOf2(c)) {
+            switch (binary->op) {
+              case MulInt64:
+                return optimizePowerOf2Mul(binary, c);
+              case RemUInt64:
+                return optimizePowerOf2URem(binary, c);
+              case DivUInt64:
+                return optimizePowerOf2UDiv(binary, c);
+              default:
+                break;
```

Comment:
Since the i32 and i64 case share so much code, I think it would make sense to combine them. The only thing that would be conditional would be the `value.geti32()` vs `value.geti64()`.

Reply:
`optimizePowerOf2Mul` and etc are generic methods. So for merge this branches I should refactor them to runtime parametric and add `type` argument (same as Abtract::* methods).

Reply:
So it will looks like this:
```cpp
if (right->type == Type::i32 || right->type == Type::i64) {
   auto c = right->value.getInteger();
   if (IsPowerOf2(c, right->type)) {
     switch (binary->op) {
        case MulInt32:
          return optimizePowerOf2Mul(binary, c, right->type);
        case RemUInt32:
          return optimizePowerOf2URem(binary, c, right->type);
        case DivUInt32:
          return optimizePowerOf2UDiv(binary, c, right->type);
        default:
          break;
      }
   }
}
```

Reply:
Another way is refactor switch part and move this to separate generic function:
```cpp
template<typename T>
wasm::Expression* optimizePowerOf2Binary(wasm::Binary* binary, T c) {
    if (IsPowerOf2<T>(c)) {
         switch (binary->op) {
              case MulInt32:
                return optimizePowerOf2Mul<T>(binary, c);
              case RemUInt32:
                return optimizePowerOf2URem<T>(binary, c);
              case DivUInt32:
                return optimizePowerOf2UDiv<T>(binary, c);
              default:
                assert(0);
            }
     }
}
```
and
```cpp
if (right->type == Type::i32) {
   return optimizePowerOf2Binary(binary, right->value.geti32(););
}
if (right->type == Type::i64) {
   return optimizePowerOf2Binary(binary, right->value.geti64());
}
```

Reply:
Ohh, I forgot it wouldn't type check if you mixed them that way. In that case, this LGTM as-is.